engine/lib/filestore.php

```diff
@@ -598,7 +609,174 @@ function elgg_get_inline_url(\ElggFile $file, $use_cookie = false, $expires = fa
 	return $file_svc->getURL();
 }
 
+/**
+ * Returns a file's URL suitable for embedding in a text editor
+ * We can not use elgg_get_inline_url() for these purposes due to a URL structure bound to user session and file modification time
+ * This function returns a generic (permanent) URL that will then be resolved to an inline URL whenever requested.
+ * 
+ * @param \ElggFile $file File entity (must have a guid)
+ * @param string    $size Thumnail size
+ * @return string
+ * @since 2.2
+ */
+function elgg_get_embed_url(\ElggFile $file, $size) {
+	return "embed-file/$file->guid/$size";
+}
+
+/**
+ * Create thumbnails for an image file
+ *
+ * @param \ElggFile $file File entity
+ * @return bool
+ */
+function elgg_create_thumbnails(\ElggFile $file) {
+
+	$prefix = 'file/';
+	$filestorename = elgg_substr($file->getFilename(), elgg_strlen($prefix));
+	
+	$sizes = [
+		'small' => [
+			'w' => 60,
+			'square' => true,
+			'metadata_name' => 'thumbnail',
+			'filename_prefix' => 'thumb',
+		],
+		'medium' => [
+			'w' => 153,
+			'square' => true,
+			'metadata_name' => 'smallthumb',
+			'filename_prefix' => 'smallthumb',
+		],
+		'large' => [
+			'w' => 600,
+			'square' => false,
+			'metadata_name' => 'largethumb',
+			'filename_prefix' => 'largethumb',
+		],
+	];
+
+	foreach ($sizes as $size => $opts) {
+		$metadata_name = $opts['metadata_name'];
+		$filename_prefix = $opts['filename_prefix'];
+		$width = $opts['w'];
+		$square = $opts['square'];
+		if (!isset($file->$metadata_name)) {
+			$file->$metadata_name = $prefix . $filename_prefix . $filestorename;
+		}
+		$thumb = elgg_get_thumbnail($file, $size);
+		$thumbnail = get_resized_image_from_existing_file($file->getFilenameOnFilestore(), $width, $width, $square);
+		if ($thumbnail) {
+			$thumb->open("write");
+			$thumb->write($thumbnail);
+			$thumb->close();
+			unset($thumbnail);
+		} else {
+			elgg_clear_thumbnails($file);
+			return false;
+		}
+	}
+	
+	$file->icontime = time();
+	return true;
+
+}
+
+	/**
+ * Clear file entity's thumbnails
+ *
+ * @param \ElggFile $file File entity
+ * @return void
+ */
+function elgg_clear_thumbnails(\ElggFile $file) {
+
+	foreach (['small', 'medium', 'large'] as $size) {
+		$thumb = elgg_get_thumbnail($file, $size);
+		$thumb->delete();
+	}
+
+	unset($file->icontime);
+	unset($file->thumbnail);
+	unset($file->smallthumb);
+	unset($file->largethumb);
+}
+
+/**
+ * Returns a file that represents file entity's thumbnail on filestore
+ * 
+ * @param \ElggFile $file File entity (must have a guid)
+ * @param string    $size Thumnail size
+ * @return \ElggFile
+ * @since 2.2
+ */
+function elgg_get_thumbnail(\ElggFile $file, $size = 'medium') {
```

Comment:
We can use it later to create missing thumbnails. At this point, I just wanted to remove all the boilerplate.